### PR TITLE
Gun Overlays are now ∞²% more efficient 

### PR DESCRIPTION
**Pull request by Robustin**
Nerds: "Sure Robustin, you improved atmos, but thats nothing compared to overlay shit hitting the fan on a populated server"

Me: Hold my scraps

------------

Self-charging guns were doing a shitload of work refreshing their overlay every few seconds of their entire existence. Turns out the floral somatoray doesn't need to be sitting there redrawing its overlay thousands of times per round as it sits on a table. 

Also guns only change overlays if their "ratio" of current energy/max energy hits a certain breakpoint (i.e. eguns hold 10 "shots" but only 4 bars to show your ammo). By simply storing the ratio from the previous update_icon() we can decide if we want to go to all the trouble of redrawing the overlay from scratch.

Lastly KA's are a meme and use a totally different ammo/icon system, yet for some baffling reason they still called their parent (egun) update_icon which would do a bunch of complex but meaningless overlay calculations before jumping to the KA proc and going "lol JK here's the actual overlay". 

**Edit: Oh god I figured out why KA's called eguns update, because it needed eguns to call guns... and it needed guns because EVERY FUCKING TIME ANY WEAPON WITH AN ATTACHMENT FIRED, RECHARGED, OR GOT BREATHED ON IT WOULD DELETE AND REBUILD THE ATTACHMENT OVERLAY FROM SCRATCH AHHHHHH... so I fixed that and now guns are ∞² more efficient, see the new table:**

  | Spawned 5 Floral Somatorays and   left on ground for 10m
-- | --
Current | 1817ms
Box of Scraps | 9ms

  | Shot a KA w/attachments 20 times
-- | --
Current | 139ms
Box of Scraps | 3ms

### CODE REVIEW

**Comment by Tlaltecuhtli** [comment]
where are the cool graphs

**Review by Tlaltecuhtli — CHANGES_REQUESTED** [review]
add cool graphs to op or no merge

**Comment by MrDoomBringer** [comment]
image

**Comment by Kromgar** [comment]
image

We're hitting warp speed men activate the gellar field or spacetime might go backwards

**Comment by Robustin** [comment]
I got the attachment situation deal with and posted some numbers

**Comment by Bombpants** [comment]
How much do all your box of scraps performance improvements actually improve performance? I am not familiar at all with how the server (or client or crystal ball) processes stuff.  

Will 24 hour rounds be possible without everything lagging to shit?

**Comment by Robustin** [comment]
Maybe someday. I'm just doing stuff one at a time and I doubt the changes so far will suddenly make long rounds lag-free.

My first atmos PR (react) was geared toward roundstart "default" behavior. It's the sort of thing that helps more when the station is close to its "default" state and less when its burning down from plasma fire. Other stuff like this PR will actually help more as the station gets busier since you will have more guns processing and coming into existence as the round drags on. From what I can see in the profiles I've looked at for Bagil, the biggest lag source is human icon updates, as I'm learning more about icons and overlays in these smaller PR's hopefully I'll understand it well enough to make a decisive improvement.

Maybe a maintainer can give you a better answer but it's all very uncertain for now, these tests are under "ideal" conditions and even horribly wasteful stuff like self-recharging guns getting fixed is a drop in the bucket (0.5-2%) of what human icons on a busy server are using up.

**Review by Tlaltecuhtli — APPROVED** [review]
very simple graph, doesnt include a wizard summoning guns x5 in an empty server filled with monkey men 6/10

**Comment by Robustin** [comment]
All I can say is we're lucky that nobody made a self-charging gun that could hold attachments before this PR.

**Comment by GuppyLaxx** [comment]
cant you put a seclight on an AEG

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/energy/special.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/projectiles/guns/energy/special.dm]: Consider using `empty_state` instead of overhauling the proc so it can inherit.

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: If you set an empty state for everything you can forgo the check (further optimizing)

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
for `/obj/item/gun/energy/kinetic_accelerator` 's `Initialize`, consider the following:

`if(isnull(empty_state))`
`empty_state = "[icon_state]_empty"`

This will store the icon and will not require the crossbow to constantly check the `icon_state` when it needs to update, as well as allowing inheritance which you can further optimize as listed below :)

File: code/modules/projectiles/guns/energy/special.dm
Comment (Robustin) [reply code/modules/projectiles/guns/energy/special.dm]: Its just awkward, ebows were made a child of KA's but they still use the traditional _empty icon setting.

Part of me thinks I should just kill empty_state since we only actually use it for the KA and it's just a weird exception.

Comment (ExcessiveUseOfCobblestone) [reply code/modules/projectiles/guns/energy/special.dm]: I would kill it and use this proc as the KA update_icon (which would inherit into ebows).

The KA uses _empty too, i'm not sure why they bothered to make it a var unless that was an attempt at optimization.

Comment (Robustin) [reply code/modules/projectiles/guns/energy/special.dm]: It probably was but it definitely never caught on.

**Comment by optimumtact** [comment]
why is it so hard to make a change that compiles and then wait for someone to merge it

**Comment by Robustin** [comment]
@optimumtact 

Why is it so hard to reopen a PR when I told you that it compiles now.

**Comment by kevinz000** [comment]
Lol

File: code/modules/projectiles/gun.dm
Comment (AnturK) [inline code/modules/projectiles/gun.dm]: Should still call parent.

File: code/modules/projectiles/guns/energy.dm
Comment (AnturK) [inline code/modules/projectiles/guns/energy.dm]: There's nothing stopping energy guns having bayonets/flashlights is there ?

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm
Comment (AnturK) [inline code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: Same here. These changes are not very reasonable long term.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/energy.dm
Comment (Robustin) [reply code/modules/projectiles/guns/energy.dm]: Correct but that's now handled in the attachment proc for guns

File: code/modules/projectiles/guns/energy/kinetic_accelerator.dm
Comment (Robustin) [reply code/modules/projectiles/guns/energy/kinetic_accelerator.dm]: That's fine, I just felt our gun code had gone in this direction where the behavior is so diverse across weapons that there will never be a "universal" icon change and if it is it should be done in something like gun attachments where its handled only when it needs to be handled instead of making every gun check/rebuild its attachments every time it shoots/recharges/etc.

File: code/modules/projectiles/gun.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/projectiles/gun.dm]: gun code.

File: code/modules/projectiles/guns/energy.dm
Comment (AnturK) [inline code/modules/projectiles/guns/energy.dm]: This means changes that do not update cell fail to happen ? For example update_icon after changing stun/kill modes.

**Review by AnturK — CHANGES_REQUESTED** [review]
Sorry i left this lying around so long.

**Comment by KorPhaeron** [comment]
Robustin hasnt been around lately, ping me when you're back

**Pull request closed without merging**
